lib/nib/history/compose.rb

```diff
@@ -0,0 +1,45 @@
+require 'tempfile'
+require 'tmpdir'
+
+class Nib::History::Compose
+  attr_reader :dir, :volume_name
+
+  def initialize
+    @volume_name = 'nib_history'
+    @dir = "#{Dir.tmpdir}/#{Dir.pwd.split('/').last}"
+
+    FileUtils.mkdir_p(dir)
+  end
+
+  def path
+    file.path
+  end
+
+  private
+
+  def config
+    @config ||= YAML.safe_load(`docker-compose config`)
+  end
+
+  def file
+    @file ||= Tempfile.open('compose', dir) do |compose|
+      compose.write(extended.to_yaml)
+      compose
+    end
+  end
+
+  def extended
+    config.dup.tap do |h|
```

Comment:
@binarymason - I tend to agree that this method is tough to understand. In order to make what's happening more clear I think the solution is to break the functionality down into small single responsibility objects/classes that can be composed into the desired `Hash`/`YAML`. I've taken a crack at that and have a (working) solution but I don't want to commit it just yet. I'd like to be able to compare the existing gnarly version the new object version. With that in mind here's the proposed version. There are many more lines of code but all of it is much more clear (IMO).

```ruby
require 'tempfile'
require 'tmpdir'

class Nib::History::Compose
  attr_reader :dir, :volume_name

  def initialize
    @volume_name = 'nib_history'
    @dir = "#{Dir.tmpdir}/#{Dir.pwd.split('/').last}"

    FileUtils.mkdir_p(dir)
  end

  def path
    file.path
  end

  def config
    original_config
      .merge('services' => services_config)
      .merge('volumes' => volumes_config)
  end

  private

  def original_config
    @original_config ||= YAML.safe_load(`docker-compose config`)
  end

  def file
    @file ||= Tempfile.open('compose', dir) do |compose|
      compose.write(config.to_yaml)
      compose
    end
  end

  def services_config
    Services.new(volume_name, original_config['services']).config
  end

  def volumes_config
    Volumes.new(volume_name, original_config['volumes']).config
  end

  class Services
    attr_reader :original_config, :volume_name

    def initialize(volume_name, original_config)
      @original_config = original_config
      @volume_name = volume_name
    end

    def config
      original_config.each_with_object({}) do |(name, definition), extended|
        extended[name] = Service.new(volume_name, definition).config
      end
    end
  end

  class Service
    attr_reader :original_config, :volume_name

    def initialize(volume_name, original_config)
      @original_config = original_config
      @volume_name = volume_name
    end

    def config
      original_config.merge('volumes' => volumes_config << history_config)
    end

    private

    def volumes_config
      original_config['volumes'] || []
    end

    def history_config
      "#{volume_name}:/usr/local/history"
    end
  end

  class Volumes
    attr_reader :original_config, :volume_name

    def initialize(volume_name, original_config = nil)
      @original_config = original_config || {}
      @volume_name = volume_name
    end

    def config
      original_config.merge(volume_name => nil)
    end
  end
end
```